Refactor CPluginDirectory into separatly usable classes

Montellese opened the pull request:
## Description
This PR splits up `CPluginDirectory` into the following classes:
* `CRunningScriptObserver`: Helper to track whether a script is still running or not.
* `CScriptRunner`: abstract base class to execute / run a script from an add-on synchronously or asynchronously. For synchronous execution it uses `CRunningScriptObserver` to wait for the script to finish.
* `CRunningScriptsHandler<>`: Template specialization of `CScriptRunner` with the added functionality to track running scripts and their state of a specific type. Right now this is only used by `CPluginDirectory` but in my media import work this is also used by the code that runs the media importer add-ons.

## Motivation and Context
My main goal was to extract the logic from `CPluginDirectory` to run add-on scripts and get a result back. `CPluginDirectory` now only contains plugin specific logic (i.e. processing a list of items). It uses `CRunningScriptsHandler<>` to execute the plugin scripts and wait for the result.

@enen92 asked me to PR this refactoring early for v20 because he wants to tackle the busy dialog issue and my work provides a decent cleanup of the existing code.

## How Has This Been Tested?
These commits have been part of my media import work for almost four years without any issues.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

Montellese committed:
factor out CRunningScriptsHandler from CPluginDirectory

Montellese committed:
factor out CRunningScriptObserver from CPluginDirectory

Montellese committed:
[filesystem] CPluginDirectory: remove unused retrievingDir parameter(s)

Montellese committed:
factor out CScriptRunner from CPluginDirectory

enen92 reviewed (approved):
It'd be nice to have also someone else looking into this. I've runtime tested this as part of mediaimporter for a long time and works as expected.
A few comments that may or may not apply:

- I am not sure generic is the best "name" for these interfaces as it only includes interfaces targeting script handling. Maybe interfaces/scripts or interfaces/languageinvoking. I'm thinking about an hypothetical future situation in which we have "languageinvoker" based binary-addons.

- You have some behavior in header files (RunningScriptsHandler.h and ScriptRunner.h. Not sure if this is bydesign or if they should go to the implementation/cpp file

- Doxygen signatures are missing for most of the methods, but weren't there before. Not sure if you want to be bothered with it :) I can help with that if needed.

Thanks much for the cleanup and for PRing this early

garbear commented:
A welcome change, and I see no problems here. My only advice would be to depend on CThread as a private member variable instead of via inheritance, because inheriting from CThread is an antipattern: it entangles the threading dependency with the inheritance hierarchy.

Montellese commented:
I have made the following changes:
* moved the inline implementations from `ScriptRunner.h` to `ScriptRunner.cpp`
* I have changed `CRunningScriptObserver` to implement `IRunnable` and start the thread on a member variable using a `CEvent` to handle aborting and sleeping.

fuzzard reviewed (commented):
Looking at includes, can the following be removed?

https://github.com/xbmc/xbmc/blob/da06165b1a93d51abccae87778e734d6b97d77f7/xbmc/filesystem/PluginDirectory.cpp#L29

Looks good otherwise, nice work.

Montellese commented:
> Looking at includes, can the following be removed?
> 
> https://github.com/xbmc/xbmc/blob/da06165b1a93d51abccae87778e734d6b97d77f7/xbmc/filesystem/PluginDirectory.cpp#L29
> 
> Looks good otherwise, nice work.

Thanks for the hint. There were quite a few other includes in `CPluginDirectory` which weren't needed anymore either. I've tried to clean them up.

Montellese commented:
Something seems off with the jenkins build :-/

enen92 commented:
jenkins build this please

garbear commented on xbmc/interfaces/generic/RunningScriptObserver.cpp:
```diff
@@ -0,0 +1,37 @@
+      m_event.Set();
+      break;
+    }
+  } while (!m_stopEvent.WaitMSec(20));
```
50 wakeups a second is a bit intense. why not wait between 100ms and 500ms?

fuzzard replied:
It's only a move of the existing behaviour. If you want to change that, probably best to split that change off in another commit

Montellese replied:
Yup I only copied / moved this behaviour from `CPluginDirectory` to `CRunningScriptObserver`. I only want to refactor `CPluginDirectory` into usable classes and I don't want to change any behaviour.

Montellese commented:
jenkins build this please

enen92 commented:
jenkins build this please

Montellese commented:
Looks like jenkins is finally happy. Can this be merged?

Acidzero2020 commented:
Hi @Montellese,

when i try to play a video via inputstream (amazon plugin) then it immediately quits kodi.
As soon as I revert the PR from my own build it works again.

Here a Log of your Nightly Build 08.03
https://pastebin.com/5bzp5aA8
Logcat
https://pastebin.com/CL77ndhM

Montellese commented:
@Acidzero2020 thank for the report and investigation. What's the easiest way to test `inputstream.adaptive`? I've never used it before...

Montellese commented:
@Acidzero2020 I tried to reproduce the issue using `plugin.video.youtube` with `inputstream.adaptive` enabled but all videos I tried played perfectly fine without any crashes or anything. Please also try latest master with #19416 merged. I tried with and without it and it made no difference for video playback (but browsing the plugin was much faster).

Acidzero2020 commented:
Thanks and sorry for the late reply. I have already tried this PR, browsing the addon runs much faster, unfortunately kodi still crashes directly. I've only tried it on Android Shield so far, I'll see if there's an easy way to simulate the error and test it on linux, unfortunately I need a bit of time.

Montellese commented:
Some questions to help narrow down the issue:
* Does it only happen with Amazon VOD plugin using `inputstream.adaptive`?
* Does it work without `inputstream.adaptive`?
* Can you try to add one of the videos that crashes Kodi on the Shield as a favourite and start playback from the favourites menu? That way you wouldn't start it from a plugin listing.

Unfortunately I have no idea how `inputstream.adaptive` works and I don't understand how a change to `CPluginDirectory` should affect video playback. But there can always be a strange case hidden somewhere...

Acidzero2020 commented:
here are a few answers, I hope it helps.

No there are many other plugins where it used  (moste pay content), but I tried the testplugin (https://github.com/romanvm/plugin.video.example/) here  happens the same and it uses no inputstream.
But it looks like it only happen on android platform.

Sorry for the wrong/incomplete information.

The favorites link doesn't work either.
Content of favorite file:
`<favourite name="Alligator" thumb="http://www.vidsplay.com/wp-content/uploads/2017/04/alligator-screenshot.jpg">PlayMedia(&quot;plugin://plugin.video.example/?action=play&amp;video=http%3A%2F%2Fwww.vidsplay.com%2Fwpcontent%2Fuploads%2F2017%2F04%2Falligator.mp4&quot;)</favourite>
`

If I create a .strm file with the direct link from the video it works.
Content of strm file:
`http://www.vidsplay.com/wp-content/uploads/2017/04/alligator.mp4`

Maven85 commented on xbmc/interfaces/generic/RunningScriptObserver.cpp:
```diff
@@ -0,0 +1,37 @@
+
+void CRunningScriptObserver::Abort()
+{
+  m_stopEvent.Set();
```
If you replace the line with 'm_thread.StopThread ();', there are no more crashes on Android.

I think that on Android, there might be a wrong thread terminating and this causing Kodi to crash.

Montellese replied:
Thanks for the input. I think I understand why this should solve the problem. `CRunningScriptObserver` is created in `CScriptRunner::WaitOnScriptResult()` as a local variable. Right before the local variable goes out of scope it calls `CRunningScriptObserver::Abort()`. I assumed that the destructor of `CRunningScriptObserver` would implicitly stop the thread because it would also execute the destructor of `m_thread` but depending on the order of destruction `m_stopEvent` is destroyed before `m_thread` so when the next loop in `CRunningScriptObserver::Run()` is executed `m_stopEvent` has become invalid.

The question is whether calling `CRunningScriptObserver::Abort()` should be blocking or not. If it should I need to add a `m_thread.StopThread()` after the `m_stopEvent.Set()`. If it shouldn't block I need to add a destructor to `CRunningScriptObserver` which first calls `Abort()` and then `m_thread.StopThread()`. Looking at the original code of `CPluginDirectory::CScriptObserver::Abort()` it also called `StopThread()` so that's probably the proper solution.

Montellese commented:
@Acidzero2020 @Maven85 could you give c849e329233c7dc9f21ed36c625b9f80857587c8 a try?

Maven85 commented:
Sure I'll test it out.

Maven85 commented:
Now it works as expected. Many Thanks!

Montellese commented:
> @Montellese Now it works as expected. Many Thanks!

Thanks for the fast feedback. I'll PR the fix to Kodi `master` ASAP.

Montellese commented:
See #19470.
